### PR TITLE
Add exhaustion feature for failed retweet attemps

### LINKED ISSUE #28: Failed Tweets is always retried with no exhaustion limit
Based on log, it appears that DailyWatori tried to retweet the same tweet ID for...ever without exhaustion limit.

```
$ docker-compose exec app tail bot.log
[1615507200247] ERROR (daily-watori/24 on 8c1329a4e332): Failed to retweet a tweet with id 1358097637912707077
[1615507200318] INFO (daily-watori/24 on 8c1329a4e332): Fetched 9 tweets from 21 users
[1615536000002] INFO (daily-watori/24 on 8c1329a4e332): Scheduled retweet started
[1615536000208] ERROR (daily-watori/24 on 8c1329a4e332): Failed to retweet a tweet with id 1358361630724136962
[1615536000220] ERROR (daily-watori/24 on 8c1329a4e332): Failed to retweet a tweet with id 1358097637912707077
[1615536000232] ERROR (daily-watori/24 on 8c1329a4e332): Failed to retweet a tweet with id 1358097690966466561
[1615564800001] INFO (daily-watori/24 on 8c1329a4e332): Scheduled retweet started
[1615564800217] ERROR (daily-watori/24 on 8c1329a4e332): Failed to retweet a tweet with id 1358361630724136962
[1615564800234] ERROR (daily-watori/24 on 8c1329a4e332): Failed to retweet a tweet with id 1358097637912707077
[1615564800245] ERROR (daily-watori/24 on 8c1329a4e332): Failed to retweet a tweet with id 1358097690966466561
```

### PR DESCRIPTION
**Pull request by chez14**
In this PR:
- Add `retweetedAt` with type datetime.
- Add `failedAt` with type datetime.
- Retweet will mark both `hasRetweeted` and `retweetedAt`.
- `getApprovedTweets` will now only find tweets that doesnt have `retweetedAt` and `failedAt` with `approvedAt` datetime is filled.
- Docker image for Development and Production is now updated to Nodejs v14.

Will fix #28, and possibly will improve #27.

### CODE REVIEW

**Comment by chez14** [comment]
Hi @Namchee,

Can you help me review this? Does this meet your coding standard?

Thanks!

File: migrations/20210312233834-BetterFailedRetweetStatus.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { Db } from 'mongodb';
+import { TweetEntity } from '../src/entity/tweet';
+
+module.exports = {
+  async up(db: Db): Promise<void> {
+    await db.collection<TweetEntity>('tweet').updateMany({}, { $set: { failedAt: undefined, retweetedAt: undefined } });
```
Comment (Namchee) [inline migrations/20210312233834-BetterFailedRetweetStatus.ts:6]: This is confusing, could you please create a newline for each steps including the query?

File: migrations/20210312233834-BetterFailedRetweetStatus.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    await db.collection<TweetEntity>('tweet').updateMany({ hasRetweeted: true }, { $set: { retweetedAt: (new Date()) } });
```
Comment (Namchee) [inline migrations/20210312233834-BetterFailedRetweetStatus.ts:7]: Same as above

File: migrations/20210312233834-BetterFailedRetweetStatus.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  },
+
+  async down(db: Db): Promise<void> {
+    await db.collection<TweetEntity>('tweet').updateMany({}, { $unset: { failedAt: 1, retweetedAt: 1 } });
```
Comment (Namchee) [inline migrations/20210312233834-BetterFailedRetweetStatus.ts:11]: Same as above

File: migrations/20210312233834-BetterFailedRetweetStatus.ts (hunk 1 above)
Comment (Namchee) [reply migrations/20210312233834-BetterFailedRetweetStatus.ts:6]: Also, will this cause a trouble for tweets that have already been sent?

**Review by Namchee — CHANGES_REQUESTED** [review]

File: migrations/20210312233834-BetterFailedRetweetStatus.ts (hunk 1 above)
Comment (chez14) [reply migrations/20210312233834-BetterFailedRetweetStatus.ts:6]: > Also, will this cause a trouble for tweets that have already been sent?

Could you please elaborate on possible troubles that may arise? I was trying to add new column to current colletion and mark scheduled tweets with `hasRetweeted` set as true to fill the `retweetedAt` column. I'm not really familiar with Mongo, so I need your advice.

Comment (Namchee) [reply migrations/20210312233834-BetterFailedRetweetStatus.ts]: Say, I have a tweet with `{ tweetId: 'random' }` that has been retweeted already, so `retweetedAt` will be `retweetedAt: <date>`.

The bot will periodically check the database of any unretweeted tweets, which is fulfilled when `retweetedAt != null`. Therefore, the bot will attempt to retweet the said tweet, which shouldn't be possible as the account has already retweeted that.

Comment (Namchee) [reply migrations/20210312233834-BetterFailedRetweetStatus.ts]: Whoops, seems I misread the code somewhere. Accepted.

**Review by Namchee — APPROVED** [review]

**Comment by chez14** [comment]
DB backed up, time to merge!

Arigathanks @Namchee!

**Comment by repurika** [comment]
:rocket: PR was released in [`v0.3.0`](https://github.com/direbus/daily-watori/releases/tag/v0.3.0) :rocket:

**Pull request merged**